File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');
const N = Number(input.shift());
const paper = [];
for (let i = 0; i < N; i++) {
    paper.push(input.shift().split(' ').map(Number));
}

let white = 0;
let blue = 0;

function solve(x, y, size) {
    const color = paper[x][y];
    let allSame = true;
    
    for (let i = x; i < x + size; i++) {
        for (let j = y; j < y + size; j++) {
            if (paper[i][j] != color) {
                allSame = false;
                break;
            }
        }
    }
    
    if (allSame) {
        if (color === 0) {
            white++;
        } else {
            blue++;
        }
        return;
    }
    
    const half = Math.floor(size / 2);
    solve(x, y, half);
    solve(x + half, y, half);
    solve(x, y + half, half);
    solve(x + half, y + half, half);
}

solve(0, 0, N);
console.log(white);
console.log(blue);

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys
input = sys.stdin.readline

def solve(x, y, size):
    global white, blue
    color = paper[x][y]
    all_same = True
    for i in range(x, x + size):
        for j in range(y, y + size):
            if paper[i][j] != color:
                all_same = False
                break
    
    if all_same:
        if color == 0:
            white += 1
        else:
            blue += 1
        return
    
    half = size // 2
    solve(x, y, half)
    solve(x + half, y, half)
    solve(x, y + half, half)
    solve(x + half, y + half, half)

N = int(input())
paper = [list(map(int, input().split())) for _ in range(N)]
white, blue = 0, 0
solve(0, 0, N)
print(white)
print(blue)